ceph/ceph_admin/helper.py

```diff
@@ -2,8 +2,216 @@
 Contains helper functions that can used across the module.
 """
 import logging
+import tempfile
+from os.path import dirname
 
-LOG = logging.getLogger()
+from jinja2 import Template
+
+from ceph.utils import get_node_by_id, get_nodes_by_ids
+
+LOG = logging.getLogger(__name__)
+
+
+class UnknownSpecFound(Exception):
+    pass
+
+
+class GenerateServiceSpec:
+    """prepare spec file based on provided config"""
+
+    COMMON_SERVICES = [
+        "mon",
+        "mgr",
+        "alertmanager",
+        "crash",
+        "grafana",
+        "node-exporter",
+        "prometheus",
+    ]
+
+    def __init__(self, node, cluster, specs):
+        """
+        Initialize the PrepareSpec
+
+        Args:
+            node: ceph node where spec file to be created
+            cluster: ceph cluster (ceph-nodes)
+            specs: service specifications
+
+        specs:
+          - service_type: host
+            address: true
+            labels: apply-all-labels
+            nodes:
+                - node2
+                - node3
+          - service_type: mon
+            placement:
+              nodes:
+                - node2
+                - node3
+          - service_type: mgr
+            placement:
+                count: 2
+          - service_type: alertmanager
+            placement:
+                count: 1
+          - service_type: crash
+            placement:
+                host_pattern: '*'
+          - service_type: grafana
+            placement:
+                count: 1
+          - service_type: node-exporter
+            placement:
+                host_pattern: '*'
+          - service_type: prometheus
+            placement:
+                count: 1
+        """
+        self.cluster = cluster
+        self.node = node
+        self.specs = specs
+        self.template_path = dirname(__file__) + "/jinja_templates/"
+
+    @staticmethod
+    def get_hostname(node):
+        return node.shortname
+
+    @staticmethod
+    def get_addr(node):
+        return node.ip_address
+
+    @staticmethod
+    def get_labels(node):
+        return node.role.role_list
+
+    def _get_template(self, service_type):
+        """
+        Return Jinja template based on the service_type
+        Args:
+            service_type: service name (ex., "host")
+        Returns:
+            template
+        """
+        path = self.template_path + f"{service_type}.jinja"
+        with open(path) as fd:
+            template = fd.read()
+        return Template(template)
+
+    def render_hosts(self, spec):
```

Comment:
Would it make sense to rename the method to `generate_host_spec`?

Reply:
Done.